Simplify repainting logic in RenderLayer::recursiveUpdateLayerPositions()

smfr opened the pull request:
#### 9dc96cf594dd31247f37fa53d3cd000471be0882
<pre>
Simplify repainting logic in RenderLayer::recursiveUpdateLayerPositions()
<a href="https://bugs.webkit.org/show_bug.cgi?id=264461">https://bugs.webkit.org/show_bug.cgi?id=264461</a>
<a href="https://rdar.apple.com/118151709">rdar://118151709</a>

Reviewed by Alan Baradlay.

The `NeedsFullRepaint` case previously duplicate some logic that exists in `RenderElement::repaintAfterLayoutIfNeeded()`,
which is that if we&apos;re doing a full repaint, we just need to paint the clippedOverflowRects.

This logic was optimized in 270370@main to turn two repaints into one sometimes, and we can re-use it here. We know
that in the RequiresFullRepaint::Yes case, `RenderElement::repaintAfterLayoutIfNeeded()` only consults the clippedOverflowRects,
so we can fold this logic with the second `shouldRepaintAfterLayout()` clause. It&apos;s fine to pass empty oldRects if we didn&apos;t
have any.

`repaintAfterLayoutIfNeeded()` does the check for printing, so we can remove that. Also the comment duplicates comments
in that function too, so remove it.

* LayoutTests/fast/repaint/hidpi-box-with-subpixel-height-inflates-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::shouldRepaintAfterLayout const):

Canonical link: <a href="https://commits.webkit.org/270454@main">https://commits.webkit.org/270454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4147202beb7ca9b588dd5c2b9008577ae4f4cab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26772 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27600 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23366 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1524 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23526 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21981 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28181 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22930 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29035 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26862 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2667 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/922 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4048 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6127 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3122 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1037,27 +1037,14 @@ void RenderLayer::recursiveUpdateLayerPositions(RenderGeometryMap* geometryMap,
+        if (flags.contains(CheckForRepaint) && newRects) {
+            ASSERT(isSelfPaintingLayer());
+            if (shouldRepaintAfterLayout()) {
+                RenderElement::RequiresFullRepaint needsFullRepaint = m_repaintStatus == RepaintStatus::NeedsFullRepaint ? RenderElement::RequiresFullRepaint::Yes : RenderElement::RequiresFullRepaint::No;
```
auto doesn't work here?

smfr replied:
It does; fixing.

smfr committed:
Simplify repainting logic in RenderLayer::recursiveUpdateLayerPositions()
https://bugs.webkit.org/show_bug.cgi?id=264461
rdar://118151709

Reviewed by Alan Baradlay.

The `NeedsFullRepaint` case previously duplicate some logic that exists in `RenderElement::repaintAfterLayoutIfNeeded()`,
which is that if we're doing a full repaint, we just need to paint the clippedOverflowRects.

This logic was optimized in 270370@main to turn two repaints into one sometimes, and we can re-use it here. We know
that in the RequiresFullRepaint::Yes case, `RenderElement::repaintAfterLayoutIfNeeded()` only consults the clippedOverflowRects,
so we can fold this logic with the second `shouldRepaintAfterLayout()` clause. It's fine to pass empty oldRects if we didn't
have any.

`repaintAfterLayoutIfNeeded()` does the check for printing, so we can remove that. Also the comment duplicates comments
in that function too, so remove it.

* LayoutTests/fast/repaint/hidpi-box-with-subpixel-height-inflates-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::shouldRepaintAfterLayout const):

Canonical link: https://commits.webkit.org/270454@main
